src/server/verification/verificationAPI.js

```diff
@@ -353,50 +355,132 @@ const setup = (app: Router, verifier: VerificationAPI, storage: StorageAPI) => {
   )
 
   /**
-   * @api {get} /verify/w3/logintoken get W3 login token for current user
-   * @apiName Get W3 Login Token
+   * @api {get} /verify/bonuses check if there is available bonuses to charge on user's wallet and do it
+   * @apiName Web3 Charge Bonuses
    * @apiGroup Verification
    *
    * @apiSuccess {Number} ok
-   * @apiSuccess {String} loginToken
    * @ignore
    */
   app.get(
-    '/verify/w3/logintoken',
+    '/verify/bonuses',
     passport.authenticate('jwt', { session: false }),
     wrapAsync(async (req, res, next) => {
-      const { user, log } = req
-      const logger = log.child({ from: 'verificationAPI - login/token' })
+      const log = req.log.child({ from: 'verificationAPI - verify/bonuses' })
+
+      const { user: currentUser } = req
+      let wallet_token = currentUser.w3Token
 
-      let loginToken = user.loginToken
+      if (!wallet_token) {
+        const secureHash = md5(currentUser.email + conf.secure_key)
 
-      if (!loginToken) {
-        const secureHash = md5(user.email + conf.secure_key)
-        const web3Response = await fetch(`${conf.web3SiteUrl}/api/wl/user`, {
+        const w3UserRes = await fetch(`${conf.web3SiteUrl}/api/wl/user`, {
           method: 'PUT',
           headers: {
             'Content-Type': 'application/json'
           },
           body: JSON.stringify({
-            secure_hash: secureHash.toLowerCase(),
-            email: user.email
+            secure_hash: secureHash,
+            email: currentUser.email
           })
         })
           .then(res => res.json())
-          .catch(e => {
-            logger.error('Get Web3 Login Response Failed', e)
+          .catch(err => {
+            log.error('Failed to fetch w3 token from W3 api', err.message, err)
           })
 
-        const web3ResponseData = web3Response && web3Response.data
+        const resData = w3UserRes && w3UserRes.data
 
-        if (web3ResponseData && web3ResponseData.login_token) {
-          loginToken = web3ResponseData.login_token
+        if (resData && resData.wallet_token) {
+          await storage.updateUser({ identifier: currentUser.loggedInAs, w3Token: resData.wallet_token })
+
+          wallet_token = resData.wallet_token
         }
       }
 
-      res.json({
-        ok: +Boolean(loginToken),
-        loginToken
+      if (!wallet_token) {
+        return res.status(400).json({
+          ok: -1,
+          message: 'Missed W3 token'
+        })
+      }
+
+      //const { release, fail } = await txManager.lock(currentUser.gdAddress, '')
+
+      const w3UserRes = await fetch(`${conf.web3SiteUrl}/api/wl/user`, {
```

Comment:
Maybe transferred to middleware or to some helper?

Reply:
You do a lot of work with W3, could it be better to create a provider and collect the logic of work in W3 in it? May be useful to others

Reply:
good idea

Reply:
@StanislavShevchenko yeah, that good suggestion. See changes regarding your comments in a new commits.